### ISSUE TITLE
Strategy for handling different data structure formats for different controller versions

### ISSUE THREAD

**Issue by fmessmer** [ros-industrial/universal_robot#123]
See [comment](https://github.com/ros-industrial/universal_robot/pull/117#issuecomment-58496562)

**Comment by fmessmer** [comment]
#135 should fix this!

**Comment by tnl-woi** [comment]
@ipa-fxm
Although I'm happy with your proposed changes, I'm not sure if we can call this a "strategy" ;-) Known versions up to 3.0 are supported now, but that'll be only as long as UR does not release another software version that uses a different packet format with the same size as one of our supported versions.
This is probably the best thing to do at this moment, but I'm a bit worried about when it will break again (possibly in an unfixable way). Do you know if Universal Robots is aware of this problem, and if they're willing to add a version number to the packet in the future firmware versions? I think that's the only way we'll ever be able to fix this issue consistently and reliably.

In short: I'm in favour of #135 but do not think we should close #123 until a more permanent solution is implemented.

**Comment by gavanderhoorn** [comment]
@tnl-woi: I can't say I know URs official policy on this, but in the past we've always had to play _catch-up_ with them with these kind of things. See #9 and #16 fi.

The only way right now to differentiate between packets from different versions is looking at the lengths, there is no version field.

> Known versions up to 3.0 are supported now, but that'll be only as long as UR does not release another software version that uses a different packet format with the same size as one of our supported versions

So far, changes to the packets have always resulted in a change in length. So in a way, the packet length can be considered a version number, although that doesn't guard against your example.

In any case, whether we have a version number or a packet length doesn't help us that much: Polyscope updates may change the packet layout, and the parser(s) cannot do much better than perhaps skip packets with an unknown / unexpected length. In the case of the real-time interface, there is only one packet being broadcast, so skipping that isn't really an option.

**Comment by fmessmer** [comment]
@tnl-woi 
Indeed the version-awareness depends on different package lengths per version.
I could not come up with a better solution out-of-the-box, but used the "strategy" that has been used for `AdditionalInfo` before...with a bit more comments and catchs for future protocol changes...

An idea for determining - or at least specifying - the version of the controller could be to add an obligatory launch-file argument for the ur_bringup launch files e.g. `version` similar to `robot_ip`. But the possible values for such argument should be made very clear.

**Comment by tnl-woi** [comment]
Agreed that the current strategy is okay for now. And a launch file argument is still an option in case auto-detection will not be possible anymore, so there is a fallback solution ;-)

Just to clarify, packets with unexpected length are not the problem, it will only become a problem when different versions can have the same packet length. For example, a UR5 v1.8 with Euromap67 installed has a packet length of 64+12=76 bytes, while a UR5 v3.0 without Euromap67 has a packet length of 72 bytes. This is already close, but fortunately not exactly the same length, so they still can be distinguished. However, the chance that auto-detection will reliably work decreases rapidly with the amount of versions and options, so that's why I still hope for a version number in the packets.

**Comment by fmessmer** [comment]
I guess for now the launch file argument is not needed yet.
We'll rethink the version-awareness "strategy" (length-checking, version-field in protocol or additional launchfile argument) on the next protocol change :smiley:

**Comment by gavanderhoorn** [comment]
On 31-10-2014 9:29, Felix Messmer wrote:

> I guess for now the launch file argument is not needed yet.
> We'll rethink the version-awareness "strategy" (length-checking, version-field in protocol or additional launchfile argument) on the next protocol change :smiley: 

unless absolutely necessary, I think we should really try to avoid an
additional launch file argument. As it is right now, ur_driver is one of
the most user-friendly drivers: nothing to install on your controller
and your robot is (re)programmed on node (re)start. It's essentially
plug-and-play.

It would be a shame if we had to introduce things that users can get
wrong (and they will, in my experience).
